File: chapter7/cpp/chapter7.cpp

#include "chapter7.h"

template <typename T>
double average(std::vector<T> numbers)
{
  std::pair<bool, double> result = {false, 0};
  double avg = 0;
  double cnt = 0; 

  for (auto& num : numbers) {
    avg += num;
    cnt++;
  }
  return avg / cnt;
}

template <typename T>
bool balance_point(std::vector<T> &arr, unsigned *point)
{
  int left, right;
  bool result = false;

  for (int i = 0; i < arr.size() - 1; i++) {
		left = 0, right = 0;
	
    /*left sum*/
		for (int l = 0; l <= i; l++) {
      left += arr[l];
		}
    /*right sum*/
		for (int r = i + 1; r < arr.size(); r++) {
      right += arr[r];
		}
    result |= left == right;
    
    if (result)
			*point = i;
  }
  return result;
}

template <typename T>
bool balance_index(std::vector<T> &arr)
{
  unsigned index;
  balance_point(arr, &index);
  std::cout << index;
  return index;
}

template <typename T>
double taco_truck(std::vector<std::array<T, 2>> arr)
{
  double manhattanDistance;
  double result; 
  taco_truck_limits limits;
  int x, y;
  
for (int i = 0; i < arr.size(); i++) {
    x = arr[i][0];
    y = arr[i][1];

    if (i == 0) { 
      limits.minY = y;
      limits.maxY = y;
      limits.minX = x;
      limits.maxX = x;
    } else { 
      if (y < limits.minY)
        limits.minY = y;
      if (y > limits.maxY)
        limits.maxY = y;
      if (x < limits.minX)
        limits.minX = x;
      if (x > limits.maxX)
        limits.maxX = x;
    }
  }
  
  for (int x = limits.minX; x <= limits.maxX; x++) {
    for (int y = limits.minY; y <= limits.maxY; y++) {
      manhattanDistance = 0;
      for (auto& ele: arr) 
        manhattanDistance += std::abs(x - ele[0]) + std::abs(y - ele[1]);
      if ((x == limits.minX  && y == limits.minY) || (manhattanDistance < result))
        result = manhattanDistance;      
    }
  }
  return result;
}

bool binary_search(std::vector<int> arr, int target)
{
  unsigned curr_balance_point = std::floor( arr.size() / 2);
	
	if (arr.size() == 0)
    return false;
  if (target == arr[curr_balance_point])
    return true;
  else if (target < arr[curr_balance_point])
    return binary_search(std::vector<int>(arr.begin(), arr.begin() + curr_balance_point), target);  // vector copy constructor
  else if (target > arr[curr_balance_point])
    return binary_search(std::vector<int>(arr.begin() + curr_balance_point + 1, arr.end()), target);
}

const char* min_of_sorted_rotated(std::vector< const char*> strings)
{
  unsigned curr_balance_point = std::floor(strings.size() / 2);
  
  if (strings.size() ==  1)
    return strings[0];

  if (std::strcmp(strings[curr_balance_point +1 ], strings[curr_balance_point]) < 0 )
    return min_of_sorted_rotated(std::vector<const char*>(strings.begin()+curr_balance_point+1, strings.end()))  ;
  else if (std::strcmp(strings[curr_balance_point - 1], strings[curr_balance_point]) < 0)
    return min_of_sorted_rotated(std::vector<const char*>(strings.begin(), strings.begin() + curr_balance_point));
  else 
    return strings[curr_balance_point];
}

int binary_search_string(std::string str, char character)
{
  unsigned curr_balance_point = std::floor(str.size() / 2); 
  
  if (str.size() == 0)
    return -1;
  
  if(str[curr_balance_point] == character)
    return true;

  if (character < str[curr_balance_point])
    return binary_search_string(str.substr(0, curr_balance_point), character);
  if (character > str[curr_balance_point])
    return binary_search_string(str.substr(curr_balance_point + 1), character);

}

std::vector<int> remove_duplicates(std::vector<int>& collection)
{
  std::vector<int> result;
  std::set<int> sets;
  
	for (auto& ele : collection) {
    if (sets.count(ele) == 0) {
      sets.insert(ele);
      result.insert(result.begin(), ele);
    }
  return result;
  }
}

void remove_duplicates_second(std::vector<int>& collection)
{
	int index;

	/**align like members */
	for (int i = 0; i < collection.size() - 1; i++) {
		index = i;
		for (int j = i; j < collection.size(); j++) {
			if (collection[i] == collection[j]) {
				std::swap(collection[j], collection[index++]);
			}
		}
	}

	/* remove dups*/
	index = 0;
	while (index < collection.size() - 1) {
		if (collection[index] == collection[index + 1]) {
			collection.erase(collection.begin() + index + 1);
		} else {
			index++;
		}
	}
}

void remove_duplicates_third(std::vector<int>& collection) {
	int index;
	std::set<int> sets;

	for (int i = 0; i < collection.size(); i++) {
		if (sets.count(collection[i]) == 0) {
			sets.insert(collection[i]);
		} else {
			collection.erase(collection.begin() + i);
			i--;
		}
	}
}

mode_struct mode(std::vector<int> collection)
{
	std::map<int, int> histo;
	mode_struct obj = {NULL, 0};

	for (auto &ele: collection) {
		if (histo.count(ele) == 0)
			histo.insert(std::make_pair(ele, 0));
		histo.at(ele)++;
	}

	for (auto& map_meta : histo) {
		if (obj.valid == NULL) {
			obj.valid = true;
			obj.max = map_meta.first;
		}

		if (map_meta.second > obj.max)
			obj.max = map_meta.first;
	}
	return obj;
}



void arr_buffer_copy(std::vector<int> source, std::vector<int> dest, unsigned source_start_index, unsigned dest_start_index, unsigned numVals)
{
	while (numVals > 0) {
		source[source_start_index++] = dest[dest_start_index++];
		numVals--;

		if (dest_start_index >= dest.size())
			dest_start_index = 0;

		if (source_start_index >= source.size())
			source_start_index = 0;
	}
}

unsigned int smarter_sum(unsigned int num) {
	unsigned int result = 0;
	for (int i = 0; i < num; i++)
		result += i;
	return result;
}

unsigned int faster_factorial(unsigned int num) {
	unsigned int result = 0;
	for (int i = 0; i < num; i++)
		result *= i;
	return result; 
}

unsigned iFib(unsigned int num) {
	unsigned int regs[2] = { 0, 1 };
	unsigned int tmp;

	if (num == 0)
		return num;
	
	if (num == 1)
		return num;

	for (int i = 2; i <= num; i++) {
		tmp = regs[0] + regs[1];
		regs[0] = regs[1];
		regs[1] = tmp;
	}

	return regs[1];
}

unsigned int iTrib(unsigned int num) {
	unsigned int regs[3] = { 0, 0, 1 };
	unsigned int sum;

	if (num == 0)
		return regs[0];

	if (num == 1)
		return regs[1];
	
	if (num == 2)
		return regs[2];

	for (int i = 2; i <= num; i++) {
		sum = 0;
		for (int j = 3; j--; ) {
			sum += regs[j];
		}
		regs[0] = regs[1];
		regs[1] = regs[2];
		regs[2] = sum;
	}

	return regs[2];
}

double median_sorted_arrays(std::vector<int> a, std::vector<int> b ) {
	unsigned int idx_a = 0, idx_b = 0;
	unsigned int pivot = std::floor((a.size() + b.size()) / 2);
	std::vector<float> buffer;
	double i = 0;
	double avg = 0;

	while ( idx_a + idx_b <= pivot ) {
		if (idx_a >= a.size())
			buffer.push_back(b[idx_b++]);
		else if (idx_b >= b.size())
			buffer.push_back(a[idx_a++]);
		else if (a[idx_a] <= b[idx_b])
			buffer.push_back(a[idx_a++]);
		else if (b[idx_b] <= a[idx_a])
			buffer.push_back(b[idx_b++]);
		i++;
	}

	if ((a.size() + b.size()) % 2 == 0) {
		return (buffer[i - 2] + buffer[i - 1]) / 2;
	} else {
		return buffer[i-1];
	}

}

std::string time_to_english(unsigned int num)
{
	std::string msg;
	unsigned int hour, minute, next_hour;
	std::string h_curr, h_next, tail;
	minute = num % 60;
	hour = (static_cast<int>(std::floor(num / 60)) % 24 );
	
	if (hour == 0) {
		h_curr = "midnight";
		h_next = "1am";
	} else if (hour == 12) {
		h_curr = "noon";
		h_next = "1pm";
	} else {
		tail = (hour >= 12) ? "pm" : "am";
		h_curr = std::to_string(hour) + tail;
		next_hour = (hour + 1) % 24;
		tail = std::to_string(next_hour);  // reassign  
		h_next = (next_hour == 12)? "noon" : (next_hour == 0) ? "midnight": (next_hour >12) ? tail + "pm" : tail + "am";
	}

	if (minute == 15)
		msg = "quarter past ";
	else if (minute <= 30)
		msg = (minute == 30) ?  "half past" : std::to_string(minute)  + "past";
	else if (minute == 45)
		msg = "quarter till";
	else if (minute > 30)
		msg += std::to_string(60 - minute) + " " + "until";
	
	if (msg.find("past") != std::string::npos) {
			msg +=  " " + h_curr;
	} else {
		msg += " "  +  h_next;
	}
	return msg;
}

/* sum of consercutive integers

	 = (n / 2 ) *  (x[0] - x[n-1])

*/

int missing_value(std::vector<int> numbers)
{
	int missing_sum = 0;
	int curr = 0;
	int min = 0;
	int max = 0;
	int i = 0;
	
	for (i = 0; i < numbers.size(); i++) {
		curr = numbers[i];
		missing_sum += curr;

		if (i == 0) {
			min = curr;
			max = curr;
		} 
		
		if (curr < min) {
			min = curr;
		} 
		
		if (curr > max) {
			max = curr;
		}
	}

	curr = 0.5 * ((max + min) * ( (i + 2) ));
	return curr - missing_sum;
}

unsigned  int rain_terraces(std::vector<int> heights)
{
  std::vector<int> stack;
	unsigned int curr, max;
  unsigned int  sum = 0;

  for (int i = 0; i < heights.size(); i++) {
    curr = heights[i];

    if (i == 0) 
      max = curr;
		
    if (curr < max) 
      stack.push_back(max - curr);
		
    if (curr > max) {
      while (!stack.empty()) {
        sum += stack[stack.size() - 1];
        stack.pop_back();
      }
			max = curr;
    }
  }
  return sum;
}

unsigned int last_digit_a_to_b(unsigned int a, unsigned int b)
{
	unsigned int cyclic_count = 0;
	unsigned start = b % 10;
	unsigned end = NULL;
	std::vector<int> cycle_list; 

	while (1) {
		end = (int)std::pow(b, cyclic_count + 1) % 10;
		cycle_list.push_back(end);
		if (cycle_list[cycle_list.size() - 1] == start &&  (cyclic_count+1) > 1)
			break;
		cyclic_count++;
	}
	return cycle_list[a % cyclic_count];
}

bool matrix_search(std::vector<std::vector<int>> image, std::vector<std::vector<int>> sub_image)
{
	bool match_found = false;

	for (int row = 0; row < image.size(); row++) {
		for (int col = 0; col < image[row].size(); col++) {
			match_found = true;
			
			if (image[row][col] == sub_image[0][0]) { /* matching  bit at left-top pixel*/
				for (int walk_row = 0; walk_row < sub_image.size(); walk_row++) {
					if (row + walk_row >= image.size())  /* walk past row boundary*/
						break;
					for (int walk_col = 0; walk_col < sub_image[walk_row].size(); walk_col++) {
						if (col + walk_col >= image[walk_col].size())  /* walk past col boundary*/
							break;
						match_found &=  (image[row + walk_row][col + walk_col] == sub_image[walk_row][walk_col]);
					}
				}

				if (match_found) {
					return true;
				}

			}
		}
	}

	return false;
}

std::string max_subarray(std::vector<int> numbers)
{
	int start_index = 0, end_index = 0;
	int curr_sum = 0;
	int data; 
	int index = 0;
	int b0, b1;
	std::string result[3] = { "[" , "", "]" };

	while (index < numbers.size()) {
		data = numbers[index];

		if (data >= curr_sum) {
			end_index = index;
			curr_sum += data;
		} else if (index + 1 < numbers.size()) {
			b0 = curr_sum + data + numbers[index + 1];
			b1 = numbers[index + 1];
			if (b1 > b0) {
				curr_sum = b1;
				index += 1;
				start_index = index;
			} else {
				curr_sum = b0;
				end_index = index;
			}
		}
		index++;
	}

	for (int i = start_index; i <= end_index; i++) {
		result[1] += std::to_string(numbers[i]) + ((i != end_index) ? ",": "");
	}

	return result[0] + result[1] + result[2];
}

 int main() {
	std::cout << max_subarray({1, 2, -4, 3, -2, 3, -1});
	 
	// std::vector<int> max_subarray(std::vector<int> numbers)
}

File: chapter7/python/chapter7.py
import math
def average(numbers):
  avg = 0
  if len(numbers):
    return None
  for element in numbers:
    avg += element
  return avg / numbers.size() 
   
def balance_point(collection):
  left = 0
  right = 0
  
  for i in range(0, len(collection) - 1 ):
    left = 0
    right = 0

    for l in range(0, i+1):
      left += collection[l]
    
    for r in range(i + 1, len(collection)):
      right += collection[r]
    
    print(left, "\t", right)
    
    if left == right:
      return True
  
  return False 

def balance_index(collection):
  left = 0
  right = 0
  
  for i in range(0, len(collection) - 1 ):
    left = 0
    right = 0

    for l in range(0, i+1):
      left += collection[l]
    for r in range(i, len(collection)):
      right += collection[r]
    if left == right:
      return i

class taco_truck_plane:
  def __init__(self):
    self.xmin = None
    self.xmax = None 
    self.ymin = None 
    self.ymax = None 

def taco_truck(coordinate_array):
  limits =  taco_truck_plane()
  manhattan_distance = None 
  min_manhattan_distance = None

  for array_element in coordinate_array:
    
    if limits.xmin == None:
      limits.xmin = array_element[0]
      limits.xmax = array_element[0]
      limits.ymin = array_element[1]
      limits.ymax = array_element[1]

    if array_element[0] < limits.xmin:
      limits.xmin = array_element[0]
    
    if array_element[0] > limits.xmax:
      limits.xmax = array_element[0]

    if array_element[1] < limits.ymin:
      limits.ymin = array_element[1]

    if array_element[1] > limits.ymax:
      limits.ymax = array_element[1]

  for x in range(limits.xmin, limits.xmax + 1):
    for y in range(limits.ymin, limits.ymax + 1):
      manhattan_distance = 0

      for element in coordinate_array:
        manhattan_distance += abs(x - element[0]) + abs(y - element[1]) 
      
      if min_manhattan_distance == None or manhattan_distance < min_manhattan_distance:
        min_manhattan_distance = manhattan_distance
        
  return min_manhattan_distance

def binary_search(sorted_array, target):
  mid = math.floor(len(sorted_array)/2)
  
  if not sorted_array:
    return False  
  
  if sorted_array[mid] == target :
    return True 
  elif target < sorted_array[mid]:
    return binary_search( sorted_array[0:mid] ,target)   
  else:
    return binary_search( sorted_array[mid+1 : : ] ,target)   

def min_sorted_rotated(string_list):
  mid = math.floor(len(string_list)/2)
  
  if (len(string_list) == 0):
    return None 

  if (len(string_list) == 1):
    return string_list[0]

  if len(string_list) == 2:
    return string_list[ +(string_list[0] > string_list[1]) ]
  
  if len(string_list) >= 3:
    if string_list[mid] < string_list[mid - 1] and string_list[mid] < string_list[mid + 1]:
      return string_list[mid]  
    elif string_list[mid - 1] < string_list[mid + 1]:
      return min_sorted_rotated(string_list[0 : mid])
    else :
      return min_sorted_rotated(string_list[mid + 1 : :])  

def binary_search_string(string_data, target_char):
    mid = math.floor(len(string_data)/2)
    if  len(string_data) == 0:
      return False 
    if string_data[mid] == target_char:
      return True 
    elif ord(target_char) < ord(string_data[mid]):
      return binary_search_string(string_data[0 : mid], target_char)
    else:
      return binary_search_string(string_data[mid + 1: : ], target_char)
    return False 

def flatten (collection, top = None):
  length = len(collection)
  element = None 

  if (length == 0):
    return None 

  for i in range (0, length): 
    element = collection[i]
    if isinstance(element, list) == True :
       flatten(element, collection)
    elif (top):
      if top == None:
        collection.append(element)
      else:
        top.append(element)
  return collection

def remove_dup (collection):
  histo = {}
  index = 0
  element = None 

  while index < len(collection):
    element = collection[index]
    if element not in histo:
      histo[element] = 0
    histo[element] += 1
    if histo[element] > 1:
      collection = collection[0 : index] + collection[index +1: :]
    else:
      index += 1
  return collection

def mode (collection):
  histo = {}
  most_frequent = {'value': None, 'count': None}

  for element in collection:
    if element not in histo:
      histo[element] = 0
    histo[element] +=1 
    if most_frequent['value'] == None or most_frequent['count'] <= histo[element] :
      most_frequent['value'] = element 
      most_frequent['count'] = histo[element]
  return most_frequent

def buffer_copy (source_arr, dest_arr, source_start_index, dest_start_index, n):
  for i in range(0, n):
    if (source_start_index >= len(source_arr)): # wrap around
      source_start_index = 0
    if (dest_start_index >= len(dest_arr)): # wrap around
      dest_start_index = 0
    dest_arr[dest_start_index] = source_arr[source_start_index]
    dest_start_index += 1
    source_start_index += 1

def iSigma(num):
  sum = 0
  for i in range(1, num + 1):
    sum += i
  return sum 

def Factorial(num):
  result = 1
  for i in range(1, num + 1):
    result *= i
  return result  

def iFib(num):
  regs = [0,1]
  for i in range (0, num):
    tmp = regs[0] + regs[1] + regs[2]
    regs[0] = regs[1] 
    regs[1] = tmp
  return regs[0]

def iTrib(num):
  regs = [0,0,1]
  for i in range (0, num):
    tmp = regs[0] + regs[1] + regs[2]
    regs[0] = regs[1] 
    regs[1] = regs[2] 
    regs[2] = tmp 
  return regs[0]

def median_sorted_arrays(a, b):
  length = (len(a) + len(b))
  mid = math.floor( length / 2 )
  buffer_array = []
  idx_a = 0
  idx_b = 0

  while (idx_a + idx_b   < mid  +(length % 2)):

    if idx_a >= len(a):
      buffer_array.append( b[idx_b] )
      idx_b = idx_b + 1

    elif idx_b >= len(b):
      buffer_array.append( a[idx_a] )
      idx_a = idx_a + 1

    elif a[idx_a] <= b[idx_b]:
      buffer_array.append( a[idx_a] )
      idx_a = idx_a + 1
    
    elif b[idx_b] <= a[idx_a]:
      buffer_array.append( b[idx_b] )
      idx_b = idx_b + 1
  
  return buffer_array[len(buffer_array) - 1] if length % 2 == 0 else ( buffer_array[len(buffer_array) - 1] +  buffer_array[len(buffer_array) - 2])/ 2  

def time_to_english (minutes):

  min_z = minutes % 60
  hour_z = math.floor(minutes / 60) % 24
  hour_z_next = (hour_z  + 1) % 24
  result = ''

  curr = {
    'hour': hour_z,
    'meridiem' : 'am' if hour_z < 12 else 'pm',
    'msg': 'midnight' if hour_z == 0 else 'noon' if  hour_z == 12  else  str(hour_z)
  }
  
  next = {
    'hour': hour_z_next,
    'meridiem' : 'am' if hour_z_next < 12 else 'pm',
    'msg': 'midnight' if hour_z_next == 0 else 'noon' if  hour_z_next == 12  else  str(hour_z_next)
  }


  if min_z == 15:
    result +=  'quarter past ' + curr['msg']
  elif min_z < 30:
    result +=  + str(min_z) + ' past ' + curr['msg']
  elif min_z == 30:
    result += 'half past ' + curr['msg']
  elif min_z > 30:
    result += str(60 -min_z)  +  ' til ' + next['msg']
  return result

def missing_value (numbers):
  min = None 
  max = None 
  length = len(numbers) + 1 

  missing_sum = 0 
  real_sum = 0

  for ele in numbers:
    if min == None:
      min = ele
      max = ele 

    if ele < min:
      min = ele 
    
    if ele > max:
      max = ele 

    missing_sum += ele 

  real_sum = 0.5 * ( (length)  * (max + min) ) 
  return  real_sum - missing_sum

def rain_terraces (terraces):
  height = 0 
  curr_sum = 0 
  terrace_height = None 
  units = 0

  for i in range(0, len(terraces)):
    terrace_height = terraces[i]

    if (terrace_height >= height):
      if curr_sum:
        units += curr_sum
        curr_sum = 0
      height = terrace_height 
    
    if (terrace_height < height):
      curr_sum += height - terrace_height 
  return units 

def last_digit_a_to_b(a,b):
  count = 0
  start = b % 10
  end = None 
  cyclic_digits = [] 

  while end != start:
    count += 1
    end  = b**(count+ 1) % 10
    cyclic_digits.append(end)

  return cyclic_digits[ a % count]


def matrix_search(image, sub_image):
  match_found = None

  for r in range(0, len(image)):
    for  c in range(0, len(image[r])):

      if image[r][c] == sub_image[0][0]:
        match_found = True 
        
        for walk_row in range(0, len(sub_image)):
          if walk_row + r >= len(image):
            break

          for walk_col in range(0, len(sub_image[walk_row])):
            if walk_col + c >= len(image[r]):
              break
            
            match_found &= image[r + walk_row][c+walk_col] == sub_image[walk_row][walk_col]

  return match_found

def max_of_subarray (collection):
  curr_sum = 0
  start = 0 
  end = 0 
  branch_0 = None 
  branch_1 = None 
  index = 0
  number = None
  
  while index < len(collection):
    number = collection[index]

    if number >= curr_sum:
      end = index
      curr_sum += number 
    
    elif  index+ 1 < len(collection):
      branch_0 = curr_sum + number + collection[index + 1]
      branch_1 = collection[index + 1]

      if branch_1 > branch_0: # non-contigious sum path larger
        curr_sum = branch_1
        index += 1
        start = index
        end = index 
      else:  # contigious sum path larger 
        curr_sum = branch_0
        end = index + 1 
        index += 2
  
    index +=1 
  
  return[start, end]

